wpilib/wpilib/pidbase.py

```diff
@@ -0,0 +1,500 @@
+        """Set the number of previous error samples to average for tolerancing. When
+        determining whether a mechanism is on target, the user may want to use a
+        rolling average of previous measurements instead of a precise position or
+        velocity. This is useful for noisy sensors which return a few erroneous
+        measurements when the mechanism is on target. However, the mechanism will
+        not register as on target for at least the specified bufLength cycles.
+
+        :param bufLength: Number of previous cycles to average.
+        :type bufLength: int
+
+        .. deprecated:: 2018.0.0
+            Use a LinearDigitalFilter as the input
+        """
+        with self.mutex:
+            self.filter = LinearDigitalFilter.movingAverage(self.origSource, bufLength)
+            self.pidInput = self.filter
+
+    def onTarget(self):
+        """Return True if the error is within the percentage of the total input
+        range, determined by setTolerance. This assumes that the maximum and
+        minimum input were set using :func:`setInput`.
+
+        :returns: True if the error is less than the tolerance
+        """
+        with self.mutex:
+            # python-specific: this is equivalent to the NullTolerance object in Java
+            raise ValueError("No tolerance value set when calling onTarget().")
+
+    def reset(self):
+        """Reset the previous error, the integral term, and disable the controller.
+
+        """
+        with self.mutex:
+            self.prevError = 0
+            self.totalError = 0
+            self.result = 0
+
+    def pidWrite(self, output):
+        """Passes the output directly to setSetpoint().
+
+        PIDControllers can be nested by passing a PIDController as another PIDController's output.
+        In that case, the output of the parent controller becomes the input (i.e., the reference) of
+        the child.
+
+        It is the caller's responsibility to put the data into a valid form for setSetpoint().
+
+        """
+        self.setSetpoint(output)
+
+    def initSendable(self, builder):
+        builder.setSmartDashboardType("PIDController")
+        builder.setSafeState(self.reset)
+        builder.addDoubleProperty("p", self.getP, self.setP)
+        builder.addDoubleProperty("i", self.getI, self.setI)
+        builder.addDoubleProperty("d", self.getD, self.setD)
+        builder.addDoubleProperty("f", self.getF, self.setF)
+        builder.addDoubleProperty("setpoint", self.getSetpoint, self.setSetpoint)
+
+    def getContinuousError(self, error):
+        """
+        Wraps error around for continuous inputs. The original error is
+        returned if continuous mode is disabled. This is an unsynchronized
+        function.
+
+        :param error: The current error of the PID controller.
+        :return: Error for continuous inputs.
+        """
+        if self.continuous and self.inputRange > 0:
+            error %= self.inputRange
```

Comment:
You know, I keep staring at this line. I know you didn't originally write this, but the meaning of `%` in Python is different to other C-like languages, so `error` is always positive here. The rough equivalent of C/Java's `%` is `math.fmod()`.

I sat down at 1am thinking this out, and I'm pretty sure the following is equivalent to both what upstream is doing and what this is doing:

```python
error %= self.inputRange
if error > self.inputRange / 2:
    return error - self.inputRange
```

Reply:
Pretty sure I looked at that and wondered about it, but figured "meh, it's in the existing codebase, it's probably fine"

Reply:
Yeah. I'm surprised we haven't ported across the PID tolerance tests.